homeassistant/components/mqtt/__init__.py

```diff
@@ -185,24 +171,22 @@ def mqtt_topic_subscriber(event):
         if not _match_topic(topic, event.data[ATTR_TOPIC]):
             return
 
-        if asyncio.iscoroutinefunction(callback):
-            yield from callback(
-                event.data[ATTR_TOPIC], event.data[ATTR_PAYLOAD],
-                event.data[ATTR_QOS])
-        else:
-            hass.add_job(callback, event.data[ATTR_TOPIC],
-                         event.data[ATTR_PAYLOAD], event.data[ATTR_QOS],
-                         priority=JobPriority.EVENT_CALLBACK)
+        hass.async_add_job(callback, event.data[ATTR_TOPIC],
+                           event.data[ATTR_PAYLOAD], event.data[ATTR_QOS])
 
     async_remove = hass.bus.async_listen(EVENT_MQTT_MESSAGE_RECEIVED,
                                          mqtt_topic_subscriber)
 
     # Future: track subscriber count and unsubscribe in remove
-    MQTT_CLIENT.subscribe(topic, qos)
+    hass.add_job(MQTT_CLIENT.subscribe, topic, qos)
```

Comment:
Seems the MQTT_CLIENT is using non-blocking sockets, so this should be fine on the event loop.